### ISSUE TITLE
Add feature: Timesheet / Time Tracking with Kimai/Harvest or similar

### ISSUE THREAD

**Issue by xet7** [wekan/wekan#812]
Integrations, only those checked ones exist currently:

- [x] [Titra](https://github.com/wekan/wekan/wiki/Time-Tracking) GPLv3
- [ ] [Kimai](http://www.kimai.org) PHP/MySQL [MIT](https://github.com/kevinpapst/kimai2)
- [ ] Harvest

_From @xet7 on January 10, 2017 18:12_

Originally from Felipe Silveira at original Wekan forums:
https://discuss.wekan.io/t/does-wekan-offer-any-timesheet-functionality-such-as-trello-harvest/393

- track time spent on card
- total amount of hours spent by resource on each type of tasks
- create reports of useful metrics

Please add info if you know other Open Source time tracking. I did not find anything from Sandstorm App Market.

This is related to #43 and #62 and making Wekan programmable.

_Copied from original issue: wefork/wekan#64_

**Comment by xet7** [comment]
Related feature request from @BaobabCoder

Instead to see time spent on a card, see estimation of time that will be spent (in hours).
Similar feature is in Visual Studio Team Services:
https://www.visualstudio.com/en-us/docs/work/kanban/_img/filter-kb-filter-text-web.png

**Comment by xet7** [comment]
Maybe pomodoro timer like this at kanbanflow?

![pomorodo_timer_at_kanbanflow](https://cloud.githubusercontent.com/assets/15545/22906293/8231c028-f24c-11e6-9934-388baa8653a7.jpg)

**Comment by jLouzado** [comment]
http://blog.capterra.com/the-top-8-free-and-open-source-time-clock-software-systems/

[Anuko](https://www.anuko.com/time_tracker/index.htm) and [Kimai](http://www.kimai.org) seem to come up again and again.

I couldn't find API documentation for either of those though, but hopefully they have some framework to support it.

**Comment by killua-eu** [comment]
Added a bounty on this issue. Please see https://www.bountysource.com/teams/wekan/issues ...

**Comment by jasonandrewduncan** [comment]
Personal opinion; we're looking for a non-proprietary Kanban board to manage a team (my organization) across multiple projects. A must have it not a tomato timer (though I use then personally) nor a timer (time spent) but something simple as a standard since the BOT (beginning of time) simple time estimates and time spend (2 text boxes).. al-la Jira, TFS, Bugzilla... I'm new to Meteor but hope to look at this project over the next few weeks and looking to see if anyone would agree that something simpler would be a good first cut. I can just fork otherwise.

**Comment by xet7** [comment]
Here was this comment, but I do not see it, maybe it was removed?

> Personal opinion; we're looking for a non-proprietary Kanban board to manage a team (my organization) across multiple projects. A must have it not a tomato timer (though I use then personally) nor a timer (time spent) but something simple as a standard since the BOT (beginning of time) simple time estimates and time spend (2 text boxes).. al-la Jira, TFS, Bugzilla... I'm new to Meteor but hope to look at this project over the next few weeks and looking to see if anyone would agree that something simpler would be a good first cut. I can just fork otherwise.

Anybody is free to add any way to track time. If there is multiple ways, there can be menu about what to use. There is no need to fork Wekan. All contributions are welcome! :)

**Comment by xet7** [comment]
Moved to here from #1843

From @rjevnikar 

I've been asked to add touch times per-member/card. Since cards and boards now both have spentTime and isOvertime fields, I thought it might make sense to discuss and possibly re-think how this timing information is stored in the database.

My current thought is to split this information out into a separate collection. As it can apply to cards or boards-as-cards, I'll refer to it as a task duration.

A task duration would have:
- a unit of time (minute, hour, day, week, sprint, month, quarter, year) 
- a planned value
- an actual value
- an overtime flag
- a rate of time (total, per day, per week, per sprint, per month, per quarter, per year)
- a boardId
- a cardId
- a userId [optional]

Does this fit with the direction the project is going in and do you see any negatives to this approach?

**Comment by xet7** [comment]
@rjevnikar 

Yes you can add time tracking the way you described.

Because there could be a lot of data added with this, it would be nice to have some cleanup options for that data too.

**Comment by dev-love** [comment]
@xet7 
I would love a minimal approach like this: 
https://teamgridapp.com/platform.html —> Task Management Section

You can simply move task cards with drag-n-drop - whether you want to assign them to a colleague or you want to start the time tracking.

Is there an ETA for time tracking?

Thanks!

**Comment by xet7** [comment]
@dev-love 

I don't know, currently @rjevnikar is working on this, not me.

**Comment by Asgoret** [comment]
@rjevnikar Hi! Is there any news about time tracker?)

**Comment by arradoq** [comment]
Perhaps it could be started like track days maybe, substract end - start and show it automatically on spent times. Just an opinion

**Comment by okv** [comment]
Hello there.

Any news about adding time tracking functionality?

**Comment by saschafoerster** [comment]
Kimai has just published version 1.0 of Kimai 2 
https://www.kimai.org/blog/2019/kimai-v2-first-production-release-1-0/
Would be cool to see both open source tools more integrated. Maybe boards could be linked to projects in Kimai and then with a click in Wekan a time-slot is prepared for saving (and probably edition).

**Comment by pabloopez** [comment]
it would be great to see this two services working together!

**Comment by justinr1234** [comment]
Harvest!

**Comment by pabloopez** [comment]
It would be great also being able to use it whit Kimai! It's self-hosted, FOSS, you know :D

**Comment by raouldo** [comment]
++ Kimai & Wekan integration. Perfect match of FOSS software.

**Comment by willhseitz** [comment]
I found today that the fairly new time tracking software, [titra](https://github.com/kromitgmbh/titra/), has a Wekan integration.
https://github.com/kromitgmbh/titra/wiki/Wekan-integration

**Comment by xet7** [comment]
@willhseitz 

Thanks a lot !!!  I will add links to it to Wekan website, and create wiki page.

**Comment by xet7** [comment]
Ok I have now added it to Wekan website and ChangeLog. Thanks!!!!

**Comment by kevinpapst** [comment]
Hi guys, maintainer of Kimai here. I was just made aware about this topic. Could anyone from the team reach out to me, so we can have a chat about possible way of integration?

**Comment by xet7** [comment]
Hi @willhseitz @saschafoerster @mbrgm etc everyone.

Question to you all, that have used both Wekan and time tracking apps, how should integration between Wekan and Kimai work? What it should track, and how? Any screenshot mockups, etc?

I did also send email to @kevinpapst , and got a reply, so we are thinking about this a little.

**Comment by willhseitz** [comment]
I would certainly check out the integration titra.io has with Kimai. I would only call it a "one-way integration", as titra uses the Wekan API to query a picklist of Cards from a single Column.

An option for Kimai would be to "stay in sync" with Wekan. Depending on how the Boards, Cards, and Columns linked up to the different objects in Kimai, any changes made in Wekan would be sent to Kimai. IMO the objects that would be linked between the two apps would be:

| Wekan | Kimai | 
| :------ | :----- |
| Board | Customer | 
| Swimlane | Project | 
| Card | Activity |

This of course depends on the use of swimlanes, so it may not fit for all users. Another option could be:

| Wekan | Kimai | 
| :------ | :----- |
| ? | Customer | 
| Board | Project | 
| Card | Activity |

Lastly, something that would be very helpful would be if the description field in a Kimai time record could be synced to Wekan as comments on a Card. I think that would require outbound webhooks on Kimai's side, which I don't think is an option today, so that would need to be developed in that project.

**Comment by saschafoerster** [comment]
It's nice to see, that Wekan und Kimai, two open source projects we really like, come together. 
Still I see that so many poeple use both tools so differently that it might be difficult to find the way that works for everyone. 

For us, we use wekan not so much for project management and as a tasklist, but as a content planning tool. Cards are for us "Content", they move from left to right as they get more details, so they develop with every step to the right.  It doesn't really make sense to us, if every card has an exact timing, because every cards has about 10 steps of work that maybe only take seconds.

For us, every Board in Wekan belongs to one single customer and one single project of this one customer. For our use of Wekan, it doesn't really matter how much time a single card has been spent on, it is only about how long the board itself was open and how long we worked in this board for the project and the customer. Our team jumps every some seconds from card to card, adds some hashtag here, add some image there, moves two cards left, etc. until in the end the content was posted successfully in some Social Media channel and then reported and archived. 

So it would be helpful if there was some "start and stop"-button, that is linked directly to this one project and customer of the board, that starts always the same task-type like "community management". It should be started and stopped directly from the board without selcting manually project and customer every time. 

Another helpful thing would be an overview how long we worked already for this one project and customer in this month (as we have monthly time budgets) and maybe what time-budget is left for this month. But even Kimai2 is missing this features (as far as I know).

But this is our personal approach to Wekan. Others might use wekan-cards more as single tasks like "Find image", that task takes 15 minutes and when it is done, they move the card to the right side in "Done". This is the way like Titra works (which is not useful for us, even though they have a nice overview of the work times of the last three month per project). When it is moved to the column "Done", then the 15 minutes in the card will be collected in the timetracker.

**Comment by gatkramp** [comment]
Thank for all the effort and hard work the team and especially @xet7 puts into Wekan. It is truly awesome to work with.

Our current solution is to use Wekan as our main planning platform, but use Titra dashboard, to track the productivity of the team.

eg. In Wekan, we write what we have been doing for how long on what day in the comments of the card we are working on. Something like this:
![image](https://user-images.githubusercontent.com/8925826/87755869-1bcdff00-c808-11ea-9dc0-6e00ae31fd68.png)

We then wrote a python script, which tracks the Wekan board, looking for comments with a specific format. Once found, it then +Tracks the time in Titra, under the users name with the time spent.

This solution has made it simple for the team to track their time, but also made it visible for management to see where there is a pitfall in hours spent. This is also a oneway integration, but we did not want to move away from Wekan. We just wanted the added benifits of Titra.

PS. The python script also sends out a blast email to all users who have not logged any hours for the week on a Friday, just before close of business.

**Comment by urmel1960** [comment]
Hi there

We (a colleaque an me) are highly interested in this feature. 

Our idea of implementation is:

- KIMAI ( first choice to integrate)
- one way sync (wekan -> kimai) 
- deleting time entries on wekan card deletes running(open) time entries in kimai
- deleting closed time entries in wekan do not delete anything in kimai
- all activities (start, end, delete) recored in system activities in wekan

recommended customizable mapping per board:

wekan | kimai | type
------------ | ------------- | ------------- 
user | user | mandatory
board title  or swimlane title or custom field | customer | mandatory
title  or swimlane title or custom field | project | mandatory
board title  or swimlane title or custom field | activity | mandatory
card title and/or description and/or comment and/or time entry comment | description | mandatory
billable | billable (custom field) | option, default checked
tags | tags | option

There should be the possibility to have multiple time entries on one card

- start (date/time) (default current timestamp when creating entry)
- end (date/time) (calculated if start and hours are provided - default current timestamp when opening running time entry)
- duration/hours (calculated if start and end are provided)
- time entry comment 
- number of hours spent on card in total (calculated sum of duration of all time entries on that card) 

Trigger would be starting a time entry on wekan card. When starting a time entry on the card , a task in kimai shoul be started - stopping the time entry on card in wekan stops the corresponding time recording in kimai as well.

A rough estimate of effort/cost would be fine.

Thanks in advance 
Urmel

**Comment by xet7** [comment]
@urmel1960 

Can you provide some screenshot mockups about what it should look like at each step in Wekan and Kimai?

I'm thinking, is this about:
a) Specifically integrating with Kimai
b) General new feature in Rules, that any REST API field in any software can be synced to/from Wekan, with mapping of fields, with triggers or scheduled times

**Comment by xet7** [comment]
> A rough estimate of effort/cost for the implementation would be fine.

1000 euro, estimated implemetation time 4 months or less. Payment options at:
https://wekan.team/commercial-support/

**Comment by urmel1960** [comment]
@xet7  

Thnak you for your first response. I will do the screenshots over the weekend.

**Comment by urmel1960** [comment]
> @urmel1960
> 
> Can you provide some screenshot mockups about what it should look like at each step in Wekan and Kimai?

@xet7 

Here is a first one showing the time entries on the card. It shows a new section on the card storing the time entries. The two buttons allow a user to start and stop time recording on the card. Users are only allowed to record times if they are assigned to a card. There should also be the possibility to add an entry "by hand" (without using the buttons) in this case start date/time and stop date/time or duration in hours need to be added via time picker or keyboard - end date/time or duration should then be calculated. The description is not mandatory.
Starting a time entry with the start button should start a corresponding time entry in kimai. Stopping a time entry with the stop button stops the corresponding running time entry in kimai. Creating a time entry "by hand" creates a corresponding time entry in kimai. 
Deleting a time entry on the card will delete the corresponding time entry in kimai - updating date/time or description in a time entry on card will update the corresponding time entry in kimai. 

If kimai is triggered by adding or changing a time entry on a card depends on the settings in system administration. Description for that will follow soon ...  :-) 

![grafik](https://user-images.githubusercontent.com/144102/113699277-a1b31280-96d5-11eb-91ff-ffb1a486f0d8.png)

**Comment by urmel1960** [comment]
and here comes the admin/setup part:

In wekan administistration there should be a section to setup the necessary parameters for the kimai integration per wekan user
Once the mappings are set, changes to settings are only allowed when all time records in wekan and the corresponding ones in kimai are closed. Changes in mappings are only valid for future time entries in kimai. Needed custom fields in wekan and/or kimai need to be created before mappings are set.

Needed mappings between wekan and kimai are (table updated since last post above)

wekan | kimai | type | explanation |
------------ | ------------- | ------------- |------------- |
user | user | mandatory
board title  or swimlane title or custom field | customer | mandatory
board title  or swimlane title or custom field | project | mandatory
board title  or swimlane title or card title or custom field | activity | mandatory
time entry description (new - see mockup above)| description | option
start date/time | start date/time | mandatory |
end date time | end date/time | mandatory | 
duration | duration | mandatory | 
billable  (custom field yes/no) | billable (custom field yes/no) | option, default checked
tags | tags | option | if yes then the tags from the card are synced into kimai time entry

I hope that helps - all questions/feedback are highly welcome

**Comment by urmel1960** [comment]
@xet7 
Hi there

I meanwhile made some clarifications about the requirements for the kimai integration - see posts above. Is that ok for you or are you still missing something else?

Regards

**Comment by xet7** [comment]
@urmel1960 

Yes it's OK for now. If I have more questions, I will add then when 1) prepayment has been received and then 2) I'm in progress of implementing this.

**Comment by urmel1960** [comment]
Here is another mockup for clarification

![grafik](https://user-images.githubusercontent.com/144102/114264379-ddf2b580-99ea-11eb-9956-cd15eef8c53b.png)

**Comment by xet7** [comment]
Hi,
I received request for invoice for this, so [added this to Roadmap](https://boards.wekan.team/b/D2SzJKZDS4Z48yeQH/wekan-open-source-kanban-board-with-mit-license/EstqPusbf97zhSvR4).

**Comment by urmel1960** [comment]
@xet7 
Hi Lauri - I really hope that everything is ok with you so far? 
The work on the time recording integration has not yet started, as it looks. We would like to ask you today not to start the work yet. 
We are still convinced that an integration of wekan and a time tracking solution would be a great added value for both products, but especially for the users. We are also very interested in this.  We are happy to support with our investment the further development of a solution like wekan - based on the experiences we can make with this project - I think here especially of your very good and everlasting commitment to the users and the wekan community. We also like your constructive way of dealing with problems, questions and criticism. We would like to take this opportunity to thank you for your commitment.
However, based on our recent experiences with kimai, we are not sure if kimai is really the right product for this integration.  We would like to do some more research and look for an alternative and ask you to put this activity on the back burner for the moment. 

Best thanks and greetings

**Comment by xet7** [comment]
@urmel1960 

Yes, please take your time to research what would right product for time tracking integration!

And thanks a lot for your kind words :slightly_smiling_face:

**Comment by jflaplante** [comment]
Sorry to disturb as I'm not the one funding that development but as an interested user I'd like to ask if there as been 'any' development on this feature? 

Just wanted to add my own time tracking experience.  For the last several years I've used a paid chrome extension for Trello called Plus!   Sadly, it was dropped recently because of a change in Chrome.  The principle was simple.  Time entries were done directly in each card by simply adding a description, number of hours and a time reference (now, -1 day, -2 days, specific date, etc).  After that, the magic was in the reporting, (by user, by cards, by board, between dates, etc)  and voilà, simple but efficient.

For me, Wekan is only missing some kind of time tracking feature to be my go to everyday tool.  I was never a fan of reproducing a project's structure inside another service simply for the purpose of timekeeping.  I always felt that the project's structure was always better keep at one place in the kanban board.  Simply adding time entries to each card is simple and efficient.
 
Thank you for reading.

**Comment by xet7** [comment]
@jflaplante 

No, there has not been any development on this yet. Currently all focus in on speedups.
